pkg/proxy/healthcheck/proxier_health.go

```diff
@@ -78,77 +79,101 @@ func NewProxierHealthServer(addr string, healthTimeout time.Duration) ProxierHea
 }
 
 func newProxierHealthServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration) *proxierHealthServer {
-	hs := &proxierHealthServer{
+	return &proxierHealthServer{
 		listener:      listener,
 		httpFactory:   httpServerFactory,
 		clock:         c,
 		addr:          addr,
 		healthTimeout: healthTimeout,
+
+		lastUpdatedMap:         make(map[v1.IPFamily]time.Time),
+		oldestPendingQueuedMap: make(map[v1.IPFamily]time.Time),
+		// The node is eligible (and thus the proxy healthy) while it's starting up
+		// and until we've processed the first node event that indicates the
+		// contrary.
+		nodeEligible: true,
 	}
-	// The node is eligible (and thus the proxy healthy) while it's starting up
-	// and until we've processed the first node event that indicates the
-	// contrary.
-	hs.nodeEligible.Store(true)
-	return hs
 }
 
-// Updated indicates that kube-proxy has successfully updated its backend, so it should
-// be considered healthy now.
-func (hs *proxierHealthServer) Updated() {
-	hs.oldestPendingQueued.Store(zeroTime)
-	hs.lastUpdated.Store(hs.clock.Now())
+// Updated indicates that proxier for the given IP family has successfully updated
+// its backend, so it should be considered healthy now.
+func (hs *proxierHealthServer) Updated(ipFamily v1.IPFamily) {
+	hs.lock.Lock()
+	defer hs.lock.Unlock()
+	hs.oldestPendingQueuedMap[ipFamily] = zeroTime
+	hs.lastUpdatedMap[ipFamily] = hs.clock.Now()
 }
 
-// QueuedUpdate indicates that the proxy has received changes from the apiserver but
-// has not yet pushed them to its backend. If the proxy does not call Updated within the
-// healthTimeout time then it will be considered unhealthy.
-func (hs *proxierHealthServer) QueuedUpdate() {
-	// Set oldestPendingQueued only if it's currently zero
-	hs.oldestPendingQueued.CompareAndSwap(zeroTime, hs.clock.Now())
+// QueuedUpdate indicates that the proxier for the given IP family has received changes
+// from the APIServer but has not yet pushed them to its backend. If the proxy does not
+// call Updated within the healthTimeout time then it will be considered unhealthy.
+func (hs *proxierHealthServer) QueuedUpdate(ipFamily v1.IPFamily) {
+	hs.lock.Lock()
+	defer hs.lock.Unlock()
+	// Set oldestPendingQueuedMap only if it's currently zero
+	if val, ok := hs.oldestPendingQueuedMap[ipFamily]; ok && val == zeroTime {
+		hs.oldestPendingQueuedMap[ipFamily] = hs.clock.Now()
+	}
```

Comment:
This logic is wrong now, because we don't initialize `oldestPendingQueuedMap` to `{ v1.IPv4Protocol: zeroTime, v1.IPv6Protocol: zeroTime }`. We initialize it to `{}`. So you want:

```
// Set oldestPendingQueuedMap[ipFamily] only if it's currently unset
if _, set := hs.oldestPendingQueuedMap[ipFamily]; !set {
        hs.oldestPendingQueuedMap[ipFamily] = hs.clock.Now()
}
```

Reply:
> if _, set := hs.oldestPendingQueuedMap[ipFamily]; !set {


This feels wrong to me, this will only allow the first `QueuedUpdate()` call to set `hs.oldestPendingQueuedMap[ipFamily]` to current time, all the subsequent calls won't be updating `hs.oldestPendingQueuedMap[ipFamily]` as the key will already be part of the map and proxy will always respond unhealthy. 


I may be wrong here, but in the current code we do not initialize the `oldestPendingQueued` and being of type `atomic.Value` its zero value won't be zeroTime and `hs.oldestPendingQueued.CompareAndSwap(zeroTime, hs.clock.Now())` won't work as old value is not zeroTime
https://github.com/kubernetes/kubernetes/blob/8e8ac86cf1118ddce2693aa8481dd02dab1385e3/pkg/proxy/healthcheck/proxier_health.go#L113

I tried to play around with it here https://go.dev/play/p/q2onw6P3aZB, `CompareAndSwap(zeroTime, currentTime)` only works if we explicitly set oldestPendingQueued to zeroTime which I guess happens in the first Updated() call.

https://github.com/kubernetes/kubernetes/blob/8e8ac86cf1118ddce2693aa8481dd02dab1385e3/pkg/proxy/healthcheck/proxier_health.go#L104

Reply:
> this will only allow the first `QueuedUpdate()` call to set `hs.oldestPendingQueuedMap[ipFa`

oh, right, `Updated` needs to be updated too.

Anyway, the stuff with comparing against `zeroTime` is just because, with the old code, `oldestPendingQueued` stored a `time.Time`, not a `*time.Time`, so we couldn't store `nil` to mean "nothing is queued", so we recognized the zero time as meaning that instead. Which is kind of weird... we really should have used a pointer.

Anyway, with the map, there's no reason to worry about comparing against the zero time. You should just have a value in the map if there is an update queued, and no value in the map when there is no update queued. So `Updated` should do `delete(hs.oldestPendingQueuedMap, ipFamily)`.

Reply:
got it, thanks !